File: src/main/typescript/Demo040ToolsInTheHandsOfAi.ts
/*
 * Copyright (c) 2025. Kazimierz Pogoda / Xemantic. All rights reserved.
 *
 * This code is provided for educational purposes as part of the
 * "Agentic AI & Creative Coding" workshop.
 * Unauthorized reproduction or distribution is prohibited.
 */

import Anthropic from '@anthropic-ai/sdk';
import type { MessageParam } from '@anthropic-ai/sdk/resources/messages';

function fibonacci(n: number, a: number = 0, b: number = 1): number {
  if (n === 0) return a;
  if (n === 1) return b;
  return fibonacci(n - 1, b, a + b);
}

const anthropic = new Anthropic();
const conversation: MessageParam[] = [];

const tools = [
  {
    name: "Fibonacci",
    description: "Calculates Fibonacci number n",
    input_schema: {
      type: "object" as const,
      properties: {
        n: {
          type: "number",
          description: "The Fibonacci number to calculate"
        }
      },
      required: ["n"]
    }
  }
];

conversation.push({
  role: "user",
  content: "What's Fibonacci number 42"
});

const toolUseResponse = await anthropic.messages.create({
  model: "claude-sonnet-4-5-20250929",
  max_tokens: 1024,
  messages: conversation,
  tools: tools,
});

console.log(`Stop reason: ${toolUseResponse.stop_reason}`);

const textContent = toolUseResponse.content.find(block => block.type === 'text');
if (textContent && textContent.type === 'text') {
  console.log(textContent.text);
}

conversation.push({
  role: "assistant",
  content: toolUseResponse.content
});

for (const block of toolUseResponse.content) {
  if (block.type === 'tool_use') {
    const result = fibonacci(block.input.n);
    conversation.push({
      role: "user",
      content: [{
        type: "tool_result",
        tool_use_id: block.id,
        content: String(result)
      }]
    });
  }
}

const finalResponse = await anthropic.messages.create({
  model: "claude-sonnet-4-5-20250929",
  max_tokens: 1024,
  messages: conversation,
  tools: tools,
});

const finalText = finalResponse.content.find(block => block.type === 'text');
if (finalText && finalText.type === 'text') {
  console.log(finalText.text);
}

/*
  Note: TypeScript requires explicit type guards when working with
  union types like content blocks. We check block.type before
  accessing type-specific properties.
*/

File: src/main/python/demo_040_tools_in_the_hands_of_ai.py
"""
Copyright (c) 2025. Kazimierz Pogoda / Xemantic. All rights reserved.

This code is provided for educational purposes as part of the
"Agentic AI & Creative Coding" workshop.
Unauthorized reproduction or distribution is prohibited.

Demo 040: Tools in the Hands of AI

What you will learn?

- Context engineering: tools as a basis for agentic use cases -
  how to define a tool input schema and connect it with Python logic.
- Cognitive science: LLMs are bad at math - do math with a calculator,
  not with harnessed stochastic entropy.
- Python: tool schemas are described as JSON Schema dicts; the model
  responds with a `tool_use` block carrying parsed arguments which
  we forward to the matching Python function.
"""

import anthropic


# A plain Python function - the "tool" implementation. The LLM never
# runs this itself; it only requests that we run it on its behalf.
def fibonacci(n, a=0, b=1):
    if n == 0:
        return a
    elif n == 1:
        return b
    else:
        return fibonacci(n - 1, b, a + b)


client = anthropic.Anthropic()

# Tool declaration: name, description and a JSON Schema for the inputs.
# The model uses the description to decide *when* to call the tool and
# the schema to know *what* to pass.
tools = [
    {
        "name": "Fibonacci",
        "description": "Calculates Fibonacci number n",
        "input_schema": {
            "type": "object",
            "properties": {
                "n": {
                    "type": "integer",
                    "description": "The Fibonacci number to calculate"
                }
            },
            "required": ["n"]
        }
    }
]

conversation = []
conversation.append({"role": "user", "content": "What's Fibonacci number 42"})

# First turn: the model decides to call the Fibonacci tool instead of
# attempting (and likely failing) to compute the value itself.
tool_use_response = client.messages.create(
    model="claude-opus-4-7",
    max_tokens=1024,
    messages=conversation,
    tools=tools
)

print(f"Stop reason: {tool_use_response.stop_reason}")
print(tool_use_response.content[0].text if tool_use_response.content[0].type == "text" else "")

conversation.append({"role": "assistant", "content": tool_use_response.content})

# Execute every tool_use block the model emitted and feed the results
# back as a `tool_result` content block keyed by the original tool_use_id.
for block in tool_use_response.content:
    if block.type == "tool_use":
        tool_result = fibonacci(block.input["n"])
        conversation.append({
            "role": "user",
            "content": [{
                "type": "tool_result",
                "tool_use_id": block.id,
                "content": str(tool_result)
            }]
        })

# Second turn: now the model has the real number and can phrase the
# final natural-language answer.
final_response = client.messages.create(
    model="claude-opus-4-7",
    max_tokens=1024,
    messages=conversation,
    tools=tools
)

print(final_response.content[0].text)
